### DISCUSSION TITLE
ESP32-S3 mit MCP2515 CAN Kommunikation zu Huawei RG4850 funktioniert nicht mehr

### DISCUSSION THREAD

**Discussion by PeterDietr** [hoylabs/OpenDTU-OnBattery#1419] [Q&A]
Ich habe hier ein Problem bei dem ich aktuell nicht mehr weiter weiß, vielleicht hatte jemand von euch schon mal das gleiche Problem.
Die Open-DTU benutze ich um PV-Überschuss mittels Huawei R4850 Netzteil in einen Speicher zu laden. Seit mehreren Monaten hat das auch absolut problemlos und zuverlässig funktioniert.
Verwendete Hardware:
ESP32-S3 WROOM-1 N16R8 (16MB) und MCP2515 CAN Modul
Seit einigen Tagen funktioniert jedoch die Kommunikation mit dem Huawei Netzteil nicht mehr. 
Auffällig für mich war das die LED auf dem MCP Board früher immer leicht geflackert hat, jetzt aber konstant leuchtet. 
Fehlermöglichkeiten:
-	Huawei Netzteil. Kann ausgeschlossen werden da mit einem anderen Huawei Netzteil das gleiche Problem besteht.
-	MCP Modul
-	ESP-S3
-	Firmware / Pin Mapping. Auch hier wurde nichts geändert.
Also folgendes versucht:
1.	Das MCP Board gegen ein neues ausgetauscht. Problem besteht weiterhin
2.	OK, also ESP32-S3 defekt? Einen anderen, jedoch nicht neuen ESP32-S3 und das MCP Board aus Versuch 1. Verwendet. Problem besteht weiterhin. 
3.	Ein neuen original verpackten ESP32-S3 und neues original verpacktes MCP Board verwendet. Problem besteht immer noch.

Alle MCP Boards funktionieren an einem älteren ESP32 4MB problemlos und können mit dem Huawei kommunizieren.

Es sieht also so aus als ob alle ESP32-S3 ein Problem haben.
In der Wiki habe ich gefunden dass die ESP32-S3 N16R8 eine SPI Spannung von nur 1,8 Volt haben, das MCP Board aber auf den SPI Pin`s 3,3 Volt verwendet. Mir ist nicht klar ob 1,8 Volt auch für die nach außen geführten nutzbaren SPI Pins gelten.
Ich steh grad etwas aufm Schlauch. Zerstören die MCP Boards die ESP32-S3 N16R8, obwohl die erste Kombi mehrere Monate „überlebt“ hat?

**Comment by stefan123t** [comment]
@PeterDietr wenn die ESP32-S3 N16R8 tatsächlich nur 1,8V Spannung können dann klappt vielleicht die Kommunikation mit dem MCP2515 nicht mehr richtig. Kaputt gehen sollte da mE erst mal nichts, nur evtl. Spannung zu niedrig am ESP Ausgang.

Hast Du davor (vor den N16R8) einen anderen ESP32-S3 mit 4/8MB verwendet der evtl höhere Ausgangsspannungen unterstützt ?

Welches Board verwendest Du genau ?

**Reply by PeterDietr** [reply]
Hallo Stefan, alle von mir verwendeten Board`s sind ESP32-S3 N16R8 Dev. Kit Board`s.  Mit einem davon hat es wie gesagt etliche Wochen ohne Probleme funktioniert. 

Ich habe hier noch ein "altes" ESP32 4mb Board liegen, mit dem funktioniert die Kommunikation zum Huawei. (Kein S3 Board)

**Comment by PeterDietr** [comment]
Dies hier habe ich in der Wiki gefunden:
![grafik](https://github.com/user-attachments/assets/3078f221-fc8c-4d18-8870-a2f521db6c46)

Auf meinem ESP steht allerdings ESP32-S3 WROOM-1 N16R8  ab der jetzt auch 1,8 Volt SPI`s hat weis ich leider nicht

**Comment by Solarteur** [comment]
Hinweis: Das Thema wurde auch hier 

https://github.com/hoylabs/OpenDTU-OnBattery/discussions/1133

besprochen.

**Reply by stefan123t** [reply]
@PeterDietr wie Solarteur anderweitig beschrieben hat, betrifft die VDD_SPI 1,8V Spannung nur die GPIOs 47 & 48.

> 2) As the VDD_SPI voltage of the ESP32-S3R8V chip has been set to 1.8 V, the working voltage for GPIO47 and GPIO48 would also be 1.8 V, which is different from other GPIOs.

D.h. das Problem könnte also evtl. auch eine andere Ursache haben. 

Was steht denn in Deinen Serial Logs ?

Hast Du bei Deinem ESP32 mit 4MB eine andere Firmware Version drauf ?

**Reply by PeterDietr** [reply]
Hallo Stefan, 
ja der ESP mit 4mb hat noch eine alte Firmware (2024.06.03)

Wie kann ich Serial Logs erstellen? Hab es mit PUTTY versucht, bekomme da aber keine Verbindung.

**Comment by Githuberix** [comment]
Hallo Stefan, ich hatte ähnliche Probleme mit  ESP32-S3, MCP2515 und R4830S.

Ähnliches wie "VDD_SPI 1,8V Spannung nur die GPIOs 47 & 48" hatte ich auch im Datenblatt gelesen. Denke auch nicht, dass das der Grund ist.

Ich habe bei KlausLi https://github.com/KlausLi/Esp-HuaweiR4850-Controller unten im Bild für die Verkabelung eine rote Leitung entdeckt. Dieser geht an den Versorgungspin des TJA1050. Lt. Datenblatt ist das ein 5V Typ mit 4,75V minimaler Spannung. Die Funktion ist reine Glückssache, wenn du das an einem Esp32 mit 3,3V an den Ausgangsleitungen benutzt. Der MCP2515 kann lt. Datenblatt von 2,7 bis 5,5V betrieben werden. D.h. der Aufbau wie bei KlausLi sollte erstmal hardwaretechnisch funktionieren.

Die Brücke für die 120 Ohm an J1 könnte fehlen.

Die Pin Zuweisung hat bei mir nicht immer geklappt , wenn ich folgendes JSON benutzt hatte

```json
[
    {
        "name": "ESP32-S3 with Huawei",
        "huawei": {
            "miso": 37,
            "mosi": 35,
            "clk": 36,
            "irq": 38,
            "power": 33,
            "cs": 40
        },
        "eth": {
            "enabled": false
        }
    }
]
```

Dann habe ich mit einem alternativen SPI geschaut, weil je nach SW, die verwendet wird, der PSRAM an Pin 35, 36, 37 hängt. Mit der nachfolgenden JSON hat es bei mir auch funktioniert.

```json
[
    {
        "name": "ESP32-S3 N16R8 with Huawei",
        "huawei": {
            "power": 33,
            "mosi": 40,
            "miso": 41,
            "clk": 39,
            "cs": 42,
            "irq": 44
        },
        "eth": {
            "enabled": false
        }
    }
]
```

Vielleicht hilft dir das bei deiner Suche.

**Reply by Githuberix** [reply]
"irq": 44 entspricht dem RX und wird auch woanders benutzt. Im Überblick auf https://opendtu-onbattery.net/hardware/esp32s3_devkit/#firmware steht dort U0RXD.

Bin mir nicht sicher, ob dass dauerhaft funktioniert und ob man gegebenfalls einen anderen GPIO nehmen kann. Kann vielleicht @schlimmchen beantworten.

**Reply by schlimmchen** [reply]
Du kannst GPIO44 eines ESP32-S3 frei verwenden, **falls** du die Firmwarevariante `generic_esp32s3_usb` verwendest. Die ohne das `_usb` nutzt diesen Pin zum Empfang von Daten der seriellen Schnittstelle für die Konsolennachrichten (auch wenn in der Praxis nichts mit eingehen Zeichen passiert). Da der Pin auch vom Bootloader benutzt wird, muss man aufpassen. Allerdings ist er dann ebenso ein Input, weshalb hier keine Probleme zu erwarten sind, auch wenn das Board startet.

> Ich habe bei KlausLi https://github.com/KlausLi/Esp-HuaweiR4850-Controller unten im Bild für die Verkabelung eine rote Leitung entdeckt. Dieser geht an den Versorgungspin des TJA1050.

Der notwendige Hack am "blauen" MCP2515 Board um den Transceiver 5V zu spendieren sollte hinlänglich bekannt sein und ist auch in unserem [Wiki](https://github.com/hoylabs/OpenDTU-OnBattery/wiki/Huawei-AC-PSU#canbus-communication-with-opendtu) beschrieben.
